igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerCache.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2017, 2018, Oracle Corporation and/or its affiliates. All rights reserved.
+ *
+ * The contents of this file are subject to the Apache License Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * If a copy of the Apache License Version 2.0 was not distributed with this file,
+ * You can obtain one at https://www.apache.org/licenses/LICENSE-2.0.html
+ */
+package com.netflix.spinnaker.igor.wercker;
+
+import com.netflix.spinnaker.igor.IgorConfigurationProperties;
+import com.netflix.spinnaker.igor.wercker.model.Run;
+import com.netflix.spinnaker.kork.jedis.RedisClientDelegate;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+/**
+ * Shared cache of build details for jenkins
+ */
+@Service
+public class WerckerCache {
+
+    private static final String POLL_STAMP = "lastPollCycleTimestamp";
+    private static final String PIPELINE_ID = "pipelineId";
+    private static final String PIPELINE_NAME = "pipelineName";
+
+    private final RedisClientDelegate redisClientDelegate;
+    private final IgorConfigurationProperties igorConfigurationProperties;
+
+    @Autowired
+    public WerckerCache(RedisClientDelegate redisClientDelegate,
+            IgorConfigurationProperties igorConfigurationProperties) {
+        this.redisClientDelegate = redisClientDelegate;
+        this.igorConfigurationProperties = igorConfigurationProperties;
+    }
+
+    public void setLastPollCycleTimestamp(String master, String pipeline, Long timestamp) {
+        String key = makeKey(master, pipeline);
+        redisClientDelegate.withCommandsClient(c -> {
+            c.hset(key, POLL_STAMP, Long.toString(timestamp));
+        });
+    }
+
+    public Long getLastPollCycleTimestamp(String master, String pipeline) {
+        return redisClientDelegate.withCommandsClient(c -> {
+            String ts = c.hget(makeKey(master, pipeline), POLL_STAMP);
+            return ts == null ? null : Long.parseLong(ts);
+        });
+    }
+
+    static Run useCreatedAtIfNotStarted(Run run) {
+        if (run.getStartedAt() == null) {
+            run.setStartedAt(run.getCreatedAt());
+        }
+        return run;
+    }
+
+    static Comparator<Run> runStartedAtComparator = new Comparator<Run>() {
```

Comment:
Similar to my above comment, it should be possible to greatly simplify this using functions in the `Comparator` library.

Something like `Comparator.nullsLast({ r -> (r.getStartedAt() ?: r.getCreatedAt())?.toInstant().toEpochMilli() })`.  That would also eliminate the need for `useCreatedIfNotStarted` which has the effect of mutating the objects being compared.  Generally I'd expect a comparator to have no side effects.

Reply:
Fixed. The comparators are now constants (of Run) to be reused.